kernel-rs/src/proc.rs

```diff
@@ -494,44 +524,113 @@ impl ProcessPool {
         -1
     }
 
-    /// Wake up all processes sleeping on waitchannel.
+    /// Wake up all processes in the pool sleeping on waitchannel.
     /// Must be called without any p->lock.
-    pub fn wakeup_pool(&mut self, targetchannel: &WaitChannel) {
-        for p in self.process.iter_mut() {
+    pub fn wakeup_pool(&mut self, target: &WaitChannel) {
+        for p in self.process_pool.iter_mut() {
             p.lock.acquire();
-            if p.waitchannel == targetchannel as _ && p.state == Procstate::SLEEPING {
-                p.state = Procstate::RUNNABLE
+            if p.waitchannel == target as _ {
+                p.wakeup()
             }
             p.lock.release();
         }
     }
-    fn run_processes(&mut self, c: *mut Cpu) {
-        for p in self.process.iter_mut() {
-            p.lock.acquire();
-            if p.state == Procstate::RUNNABLE {
-                // Switch to chosen process.  It is the process's job
-                // to release its lock and then reacquire it
-                // before jumping back to us.
-                p.state = Procstate::RUNNING;
-                unsafe {
-                    (*c).proc = p;
-                    swtch(&mut (*c).scheduler, &mut p.context);
 
-                    // Process is done running for now.
-                    // It should have changed its p->state before coming back.
-                    (*c).proc = ptr::null_mut()
-                }
-            }
-            p.lock.release();
+    /// Set up first user process.
+    pub unsafe fn user_proc_init(&mut self) {
+        let mut p = match self.alloc() {
+            Ok(proc) => proc,
+            _ => ptr::null_mut(),
+        };
```

Comment:
기존의 코드는 이 시점에서 `alloc` 이 절대 실패하지 않을 것이라고 가정하고 진행합니다. 실제로 그러므로 버그가 아니긴 한데, 첫 init 시점에서 error 가 나지 않는 것을 타입으로 보장하는 것은 어렵습니다.

비록 런타임 체크가 조금 더 들어가지만, 이 코드는 부팅 시 단 한 번만 호출되므로 `expect` 를 호출하도록 합시다.
```suggestion
        let mut p = self.alloc().expect("user_proc_init");
```

Reply:
아 이게 `self.alloc()` 이 `Ok(ptr::null_mut())` 일 수도 있네요.. null 이면 무조건 `Err` 가 되도록 위에서 수정하는 게 좋겠습니다.